source/reactor/inc/reactor/EventLoop.h

```diff
@@ -0,0 +1,73 @@
+#ifndef Reactor_EventLoop_H
+#define Reactor_EventLoop_H
+
+namespace u {
+namespace reactor {
+class EventHandlers;
+
+/**
+ * The event loop is a programming construct that waits for and dispatches events in a program.
+ *
+ * It works by making a request to some "event provider" (which generally blocks the request
+ * until an event has arrived), and then it calls the relevant event handler (i.e.: dispatches
+ * the event).
+ *
+ * The event provider used by the event loop follows the file interface, which can be selected
+ * or 'polled' (the Unix system call, not actual polling). The event loop operates
+ * asynchronously with the event originator.
+ *
+ * When the event loop forms the central control flow construct of a program, as it often does,
+ * it may be termed the main loop or main event loop. This title is appropriate because such an
+ * event loop is at the highest level of control within the program.
+ */
+class EventLoop {
+public:
+    virtual ~EventLoop() = default;
+
+    /**
+     * @brief Registers a set of event handlers for the given file descriptor.
+     *
+     * Event handler for the event that occurred will be called back when the file descriptor is
+     * ready for a read/write O operation, when an error occurs or when given timeout expires.
+     *
+     * @param fd File descriptor. Handle that identifies a resource managed within an OS:
+     * network connections, open files, timers, synchronization objects, etc.
+     * @param timeout Time to wait in milliseconds (-1 to wait indefinitely) for a read/write
+     * event to occur.

```

Comment:
I think that the only thing that matters is that `indefinitely` **would not** be an option.
Justification: like any other resource in the system, the time must be restricted explicitly.
Having a global timeout is good enough and simplifies development.

Reply:
Original [article ](http://www.dre.vanderbilt.edu/~schmidt/PDF/reactor-siemens.pdf)describing reactor pattern includes timeout handlers.

Its usage is optional. It facilitates timeout handling if required.

Reply:
> Original [article ](http://www.dre.vanderbilt.edu/~schmidt/PDF/reactor-siemens.pdf)describing reactor pattern includes timeout handlers.

It's not because it was originally described like that that it's a good idea :-)

If it is something like CppCoreGuidelines or Boost, which are the result of years of iterative improvements from a large community, I'd be less inclined to go against the "rules". But a single paper from 1995 is nothing more than good inspiration.

None of the event loop approaches I know (Qt, Glib, libev, libevent, twisted, and I'm probably forgetting half a dozen) has this kind of per-file-descriptor timeout. Because it simply makes little sense. It is only useful for one case: in a request-response scenario over a non-multiplexed channel. This is a bit of a corner case, and would be much better served with a specific pattern for that scenario (which, BTW, would be easily extendible to multiplexed channels).

> Its usage is optional.

Sure, but it adds complexity to the implementation, and complexity == bugs.

I have a very strong dislike for dead code, even if it's unit tested.

> It facilitates timeout handling if required.

Again, in one specific use case, and it doesn't serve that case very well.

Reply:
I've been thinking a lot about setting specific timeouts for each file descriptor since you wrote your first comment and the more I think, the less convinced I am about its suitability. 
I got finally convinced against its use after reviewing #1435. If I had to write this API it again, I would not include timeouts.
A `Timer` wrapper class around `timerfd` would perfectly do the job.